nntrainer/layers/conv2d_layer.cpp

```diff
@@ -474,21 +431,41 @@ void Conv2DLayer::im2col(const Tensor &in, const TensorDim &kdim,
   unsigned int width = in_width + pw * 2;
   unsigned int k_height = kdim.height();
   unsigned int k_width = kdim.width();
-  unsigned int out_height = (height - k_height) / mstride[0] + 1;
-  unsigned int out_width = (width - k_width) / mstride[1] + 1;
-  unsigned int kernel_feature_size = kdim.getFeatureLen();
-
-  /// shortcut 1: if kernel is 1x1
-  // Current setup requires this to be transposed for channel mode = true
-  // if (channel_mode && k_height * k_width == 1) {
-  //   out = in;
-  //   out.reshape({channel, height * width});
-  //   return;
-  // }
+
+  /// effective kernel height considering dilation
+  unsigned int eff_k_height = (k_height - 1) * dilation[0] + 1;
+  /// effective kernel width considering dilation
+  unsigned int eff_k_width = (k_width - 1) * dilation[1] + 1;
+
+  unsigned int out_height = (height - eff_k_height) / mstride[0] + 1;
+  unsigned int out_width = (width - eff_k_width) / mstride[1] + 1;
+
+  NNTR_THROW_IF(height < eff_k_height, std::invalid_argument)
```

Comment:
let's move these checks to `initialize()`, there is no need to perform the same check with every iteration.

Reply:
This will have to move to line 126 eventually, it's more than moving this logic but needs few more checks as the function is used in two place. 
I'll do it in a separate PR instead will add a todo comment here.